crates/networking/p2p/sync.rs

```diff
@@ -0,0 +1,137 @@
+use std::sync::Arc;
+
+use ethrex_blockchain::error::ChainError;
+use ethrex_core::{
+    types::{Block, BlockHash, BlockHeader},
+    H256,
+};
+use ethrex_storage::Store;
+use tokio::{sync::Mutex, time::Instant};
+use tracing::{debug, info, warn};
+
+use crate::kademlia::KademliaTable;
+
+/// Manager in charge the sync process
+/// Only performs full-sync but will also be in charge of snap-sync in the future
+#[derive(Debug)]
+pub struct SyncManager {
+    // true: syncmode = snap, false = syncmode = full
+    #[allow(unused)]
+    snap_mode: bool,
+    peers: Arc<Mutex<KademliaTable>>,
+}
+
+impl SyncManager {
+    pub fn new(peers: Arc<Mutex<KademliaTable>>, snap_mode: bool) -> Self {
+        Self { snap_mode, peers }
+    }
+
+    /// Starts a sync cycle, updating the state with all blocks between the current head and the sync head
+    /// TODO: only uses full sync, should also process snap sync once implemented
+    pub async fn start_sync(&mut self, mut current_head: H256, sync_head: H256, store: Store) {
+        info!("Syncing from current head {current_head} to sync_head {sync_head}");
+        let start_time = Instant::now();
+        // Request all block headers between the current head and the sync head
+        // We will begin from the current head so that we download the earliest state first
+        // This step is not parallelized
+        let mut all_block_headers = vec![];
+        let mut all_block_hashes = vec![];
+        loop {
+            let peer = self.peers.lock().await.get_peer_channels().await;
+            debug!("Requesting Block Headers from {current_head}");
+            // Request Block Headers from Peer
+            if let Some(block_headers) = peer.request_block_headers(current_head).await {
+                debug!("Received {} block headers", block_headers.len());
+                let block_hashes = block_headers
+                    .iter()
+                    .map(|header| header.compute_block_hash())
+                    .collect::<Vec<_>>();
+                // Keep headers so we can process them later
+                // Discard the first header as we already have it
+                all_block_headers.extend_from_slice(&block_headers[1..]);
+                all_block_hashes.extend_from_slice(&block_hashes[1..]);
+
+                // Check if we already reached our sync head or if we need to fetch more blocks
+                if !block_hashes.contains(&sync_head) {
+                    // Update the request to fetch the next batch
+                    current_head = *block_hashes.last().unwrap();
+                } else {
+                    // No more headers to request
+                    break;
+                }
+            }
+        }
+        // We finished fetching all headers, now we can process them
+        // TODO: snap-sync: launch tasks to fetch blocks and state in parallel
```

Comment:
Let's add issues for this.

Reply:
There is already a PR in progress for snap sync #1209